core/shared/src/main/scala/zio/internal/ConcurrentWeakHashSet.scala

```diff
@@ -0,0 +1,579 @@
+package zio.internal
+
+import zio.internal.ConcurrentWeakHashSet.Ref
+
+import java.lang.ref.WeakReference
+import scala.collection.mutable
+import java.util.concurrent.locks.ReentrantLock
+import java.lang.ref.ReferenceQueue
+import java.util.concurrent.atomic.AtomicInteger
+
+private[zio] object ConcurrentWeakHashSet {
+
+  private[internal] val MaxConcurrencyLevel: Int     = 1 << 16
+  private[internal] val MaxSegmentSize: Int          = 1 << 30
+  private[internal] val DefaultInitialCapacity: Int  = 16
+  private[internal] val DefaultLoadFactor: Float     = 0.75f
+  private[internal] val DefaultConcurrencyLevel: Int = 16
+
+  /**
+   * Reference node that links elements in the set with the same hash code
+   *
+   * @param hash
+   *   hash code of the element
+   * @param element
+   *   value passed to the weak reference
+   * @param queue
+   *   reference queue that will be used to track garbage collected elements
+   * @param nextRef
+   *   next reference node in the chain
+   * @tparam V
+   *   type of the element
+   */
+  protected class Ref[V](
+    val hash: Int,
+    element: V,
+    queue: ReferenceQueue[V],
+    var nextRef: Ref[V]
+  ) extends WeakReference[V](element, queue)
+
+  /**
+   * Access options for the update operation.
+   */
+  protected object UpdateOptions extends Enumeration {
+    type UpdateOption = Value
+    val RestructureBefore, RestructureAfter, SkipIfEmpty, Resize = Value
+  }
+
+  /**
+   * Supported results of the update operation.
+   */
+  protected object UpdateResults extends Enumeration {
+    type UpdateResult = Value
+    val None, AddElement, RemoveElement = Value
+  }
+
+  /**
+   * Base class for all update operations.
+   */
+  protected abstract class UpdateOperation[V](val options: UpdateOptions.UpdateOption*) {
```

Comment:
This can be a `case class`:

```suggestion
  protected final case class UpdateOperation[V](options: UpdateOptions.UpdateOption*) {
  }
```

Reply:
`UpdateOperation` is abstract, because it's for anonymous calls in update methods, like that:

https://github.com/zio/zio/blob/7a071ca8a14cbbccbdd7ec5e08517751b80a9cac/core/shared/src/main/scala/zio/internal/ConcurrentWeakHashSet.scala#L255-L259

This approach has its origins in the Spring reference, but I think I'll fully remove this with constant args in the final impl once we'll figure out the thing with ReentrantLock, so I'll leave it for now.